streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -186,12 +190,17 @@ public void process(final K key, final V1 value) {
 
         @SuppressWarnings("unchecked")
         private void emitNonJoinedOuterRecords(final WindowStore<KeyAndJoinSide<K>, LeftOrRightValue> store) {
+            if (minTime.minTime >= maxObservedStreamTime.get() - joinAfterMs - joinBeforeMs - joinGraceMs) {
```

Comment:
Why adding both joinAfterMs and joinBeforeMs? The records expire when `window().start() + joinAfterMs + joinGraceMs` are lower than maxStreamTime. For instance, say we have a record in the shared state store with time = 1. Now a new record arrives with time = 17.

```
inputRecordTime = 17
maxObservedTime = 17
minTime = 1
window = 10  (beforeMs = 10, afterMs = 10)
grace = 5
```

Isn't the record 1 suppose to expire and be emitted because 1 + 10 (afterMs) + 5 (grace) = 16? which is lower than maxStreamTime?

With the condition you have, the minTime registered is 1, so `(1 >= 17 - 10 - 10 - 5)` is true, and thus it returns and do not emit the record 1 until another 10 ms has passed.

Reply:
It's a performance question. Without `joinBeforeMs` we only get 3K rec/sec throughput in our benchmarks, while we get 12K rec/sec throughput with `joinBeforeMs`.

Reply:
If we want to treat it as a performance/emit latency trade-off, then maybe better not piggy-back on `beforeMs` but certain value that's irrelevant to the window size?

Reply:
That a good point, and I was concerned about it, too. Using `beforeMs` might increase the latency to much, especially for large windows.

I might actually be best, to just to processing-time periodic emits, to avoid calling `store.all()` too oftern.